### ISSUE TITLE
eosio-cpp abi generation "type is invalid", eosiocpp correct. Ignoring 'key_names' and 'key_types'

### ISSUE THREAD

**Issue by qpIlIpp** [EOSIO/eosio.cdt#106]
If we will try to generate abi for that contract with eosio-cpp : 
```
#include <eosiolib/eosio.hpp>
using namespace eosio;

//@abi table tests i64
struct[[eosio::table]] tst
{
	tst() {};
	
	uint64_t id;
	vector<uint64_t> vec;
	
	uint64_t mysize() const {return vec.size();}
	EOSLIB_SERIALIZE(tst, (id)(vec))
};

typedef eosio::multi_index
<N(tests), tst,
	indexed_by<N(tstssize), eosio::const_mem_fun<tst, uint64_t, &tst::mysize>>
> tstsT;

class hello : public eosio::contract {
  public:
      hello(account_name self) : eosio::contract(self)
		, tests(self, self)
		{}
      tstsT tests;
      //abi action
      [[eosio::action]]
      void hi( tst val ) {
         print("size: ", val.vec.size());
      }
};

EOSIO_ABI( hello, (hi) )
```
with `eosio-cpp ./hello.cpp -o ./hello.wasm --abigen`
It's ABI will contains `"type": "vector<uint64_t>"` , which will cause error `Error 3015004: The type defined in the ABI is invalid` when calling `cleos push action contractname hi '[[1,[2,3]]]' -p contractname`.

The old way to compile `eosiocpp -g ./hello.abi ./hello.cpp` will create type `"uint64[]"`

Moreover, if we will look at section tables, eosiocpp will generate 
```
"tables": [{
      "name": "tests",
      "index_type": "i64",
      "key_names": [
        "id"
      ],
      "key_types": [
        "uint64"
      ],
      "type": "tst"
    }
  ]
```

and eosio-cpp will ignore `key_names` and `key_types` and generate that:

```
"tables": [
        {
            "name": "tests",
            "type": "tst",
            "index_type": "i64",
            "key_names": [],
            "key_types": []
        }
    ],
```

**Comment by lonkly** [comment]
Duplicate of https://github.com/EOSIO/eosio.cdt/issues/109
bug with vectors abi gen

**Comment by qpIlIpp** [comment]
What did u mean? U meant that #109 is duplicate of #106? If I'm not wrong, 109 is greater than 106.

**Comment by cppfuns** [comment]
I also encountered this situation, how to resolve it? help me!!!

**Comment by qpIlIpp** [comment]
> I also encountered this situation, how to resolve it? help me!!!

The easiest way is to not use eosio-cpp, just compile with old eosiocpp .

More complex way is to to fix your abi manualy: replace vectors with arrays and add key_names and key_types.

**Comment by cppfuns** [comment]
I can't describe the mood at the moment in words, I hope I can give priority to this problem.

**Comment by cppfuns** [comment]
@larryk85

**Comment by larryk85** [comment]
The type is invalid issue with vectors has been resolved in v1.3.0.  As for key_names and key_types, they are ignored, as they are a deprecated part of the abi, these fields are a hold over from a different system for secondary tables and no longer make any sense.
